lib/mix/lib/mix/release.ex

```diff
@@ -350,6 +353,16 @@ defmodule Mix.Release do
       Mix.raise("The :steps option must contain the atom :assemble once, got: #{inspect(steps)}")
     end
 
+    Enum.reduce_while(steps, nil, fn
+      :assemble, _ -> {:halt, nil}
+      :tar, _ -> Mix.raise("The :tar step must come after :assemble")
+      _, _ -> {:cont, nil}
+    end)
```

Comment:
Can also be written as:

```elixir
if :assemble in  Enum.drop_while(steps, & &1 != :tar) do
   Mix.raise("The :tar step must come after :assemble")
end
```

Not as efficient but we probably don't care about it here.